shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-core/src/main/java/org/apache/shardingsphere/sharding/algorithm/sharding/cosid/AbstractIntervalShardingAlgorithm.java

```diff
@@ -0,0 +1,143 @@
+
+    public static final String INTERVAL_AMOUNT_KEY = "datetime-interval-amount";
+
+    public static final String ZONE_ID_KEY = "zone-id";
+
+    private volatile IntervalTimeline intervalTimeline;
+
+    private ZoneId zoneId = ZoneId.systemDefault();
+
+    @Getter
+    @Setter
+    private Properties props = new Properties();
+
+    /**
+     * get zone id.
+     *
+     * @return zone id
+     */
+    protected ZoneId getZoneId() {
+        return zoneId;
+    }
+
+    @Override
+    public void init() {
+        if (getProps().containsKey(ZONE_ID_KEY)) {
+            zoneId = ZoneId.of(getRequiredValue(ZONE_ID_KEY));
+        }
+        String logicNamePrefix = getRequiredValue(CosIdAlgorithm.LOGIC_NAME_PREFIX_KEY);
+        LocalDateTime effectiveLower = LocalDateTime.parse(getRequiredValue(DATE_TIME_LOWER_KEY), DEFAULT_DATE_TIME_FORMATTER);
+        LocalDateTime effectiveUpper = LocalDateTime.parse(getRequiredValue(DATE_TIME_UPPER_KEY), DEFAULT_DATE_TIME_FORMATTER);
+        DateTimeFormatter suffixFormatter = DateTimeFormatter.ofPattern(getRequiredValue(SHARDING_SUFFIX_FORMAT_KEY));
+        ChronoUnit stepUnit = ChronoUnit.valueOf(getRequiredValue(INTERVAL_UNIT_KEY));
+        int stepAmount = Integer.parseInt(getProps().getProperty(INTERVAL_AMOUNT_KEY, "1"));
+        intervalTimeline = new IntervalTimeline(logicNamePrefix, Range.closed(effectiveLower, effectiveUpper), IntervalStep.of(stepUnit, stepAmount), suffixFormatter);
+    }
+
+    private String getRequiredValue(final String key) {
+        return PropertiesUtil.getRequiredValue(getProps(), key);
+    }
+
+    @Override
+    public String doSharding(final Collection<String> availableTargetNames, final PreciseShardingValue<T> shardingValue) {
+        LocalDateTime shardingTime = convertShardingValue(shardingValue.getValue());
+        return this.intervalTimeline.sharding(shardingTime);
+    }
+
+    @Override
+    public Collection<String> doSharding(final Collection<String> availableTargetNames, final RangeShardingValue<T> shardingValue) {
+        Range<LocalDateTime> shardingRangeTime = convertRangeShardingValue(shardingValue.getValueRange());
+        return this.intervalTimeline.sharding(shardingRangeTime);
+    }
+
+    /**
+     * convert sharding value to {@link LocalDateTime}.
+     *
+     * @param shardingValue sharding value
+     * @return The {@link LocalDateTime} represented by the sharding value
+     */
+    protected abstract LocalDateTime convertShardingValue(T shardingValue);
+
+    private Range<LocalDateTime> convertRangeShardingValue(final Range<T> shardingValue) {
+        if (Range.all().equals(shardingValue)) {
+            return Range.all();
+        }
+        Object endpointValue = shardingValue.hasLowerBound() ? shardingValue.lowerEndpoint() : shardingValue.upperEndpoint();
+        if (endpointValue instanceof LocalDateTime) {
+            @SuppressWarnings("unchecked")
```

Comment:
@Ahoo-Wang Move this annotation to method?

Reply:
ok